src/Analyzer/Passes/CrossToInnerJoinPass.cpp

```diff
@@ -0,0 +1,254 @@
+
+    if (const auto * join_node = table_expression->as<JoinNode>())
+    {
+        return findInTableExpression(source, join_node->getLeftTableExpression())
+            || findInTableExpression(source, join_node->getRightTableExpression());
+    }
+
+    if (const auto * query_node = table_expression->as<QueryNode>())
+    {
+        return findInTableExpression(source, query_node->getJoinTree());
+    }
+
+    return false;
+}
+
+class CrossToInnerJoinVisitor : public InDepthQueryTreeVisitorWithContext<CrossToInnerJoinVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitorWithContext<CrossToInnerJoinVisitor>;
+    using Base::Base;
+
+    /// Returns false if can't rewrite cross to inner join
+    bool tryRewrite(JoinNode * join_node)
+    {
+        if (!isCrossOrComma(join_node->getKind()))
+            return true;
+
+        if (where_stack.empty())
+            return false;
+
+        auto & where_condition = *where_stack.back();
+        if (!where_condition)
+            return false;
+
+        const auto & left_table = join_node->getLeftTableExpression();
+        const auto & right_table = join_node->getRightTableExpression();
+
+        QueryTreeNodes equi_conditions;
+        QueryTreeNodes other_conditions;
+        exctractJoinConditions(where_condition, equi_conditions, other_conditions);
+        bool can_join_on_anything = false;
+        for (auto & cond : equi_conditions)
+        {
+            auto left_src = getExpressionSource(getEquiArgument(cond, 0));
+            auto right_src = getExpressionSource(getEquiArgument(cond, 1));
+            if (left_src.second && right_src.second && left_src.first && right_src.first)
+            {
+                bool can_join_on = (findInTableExpression(left_src.first, left_table) && findInTableExpression(right_src.first, right_table))
+                    || (findInTableExpression(left_src.first, right_table) && findInTableExpression(right_src.first, left_table));
+
+                if (can_join_on)
+                {
+                    can_join_on_anything = true;
+                    continue;
+                }
+            }
+
+            /// Can't join on this condition, move it to other conditions
+            other_conditions.push_back(cond);
+            cond = nullptr;
+        }
+
+        if (!can_join_on_anything)
+            return false;
+
+        equi_conditions.erase(std::remove(equi_conditions.begin(), equi_conditions.end(), nullptr), equi_conditions.end());
+        join_node->crossToInner(makeConjunction(equi_conditions));
+        where_condition = makeConjunction(other_conditions);
+        return true;
+    }
+
+    void visitImpl(QueryTreeNodePtr & node)
+    {
+        if (!isEnabled())
+            return;
+
+        if (auto * query_node = node->as<QueryNode>())
+        {
+            /// We are entering the subtree and can use WHERE condition from this subtree
+            if (auto & where_node = query_node->getWhere())
```

Comment:
It seems that we can avoid this `where_stack` completely.
Here we can extract all table expressions from query node. And then try to optimize only JOINS. Check `buildTableExpressionsStack` method from Analyzer utils.
So in visitImpl we just process each query node JOIN nodes using WHERE predicate.